consensus/src/authority/actors/backfiller/actor.rs

```diff
@@ -512,41 +476,37 @@ impl<E: Clock + GClock + Rng, C: Scheme, H: Hasher, S: Supervisor<Index = View>>
-                                if !nullifications.contains(&view) {
+                                let entry = Entry { notarization: false, view };
+                                if !self.required.contains(&entry) {
                                     debug!(view, sender = hex(&s), "unnecessary nullification");
                                     continue;
                                 }
                                 if !verify_nullification::<S,C>(&self.supervisor, &self.nullify_namespace, &nullification) {
                                     warn!(view, sender = hex(&s), "invalid nullification");
-                                    self.incorrect.insert(s.clone());
-                                    break;
+                                    self.requester.block(s.clone());
+                                    continue;
                                 }
-                                nullifications.remove(&view);
+                                self.required.remove(&entry);
                                 self.nullifications.insert(view, nullification.clone());
                                 voter.nullification(nullification).await;
                                 nullifications_found.insert(view);
                             }
 
                             // Update performance
                             if !notarizations_found.is_empty() || !nullifications_found.is_empty() {
-                                let duration = received.duration_since(status.1).unwrap();
-                                self.fetch_performance.put(s.clone(), duration);
+                                self.requester.resolve(request);
                                 debug!(
                                     notarizations_found = ?notarizations_found.into_iter().collect::<Vec<_>>(),
                                     nullifications_found = ?nullifications_found.into_iter().collect::<Vec<_>>(),
                                     sender = hex(&s),
                                     "request successful",
                                 );
+                            } else {
+                                debug!(sender = hex(&s), "response not useful");
                             }
 
                             // If still work to do, send another request
-                            if notarizations.is_empty() && nullifications.is_empty() {
-                                outstanding = None;
-                                continue;
-                            }
-                            let (notarizations, nullifications, mut sent, _) = outstanding.take().unwrap();
-                            let status = self.send_request(&notarizations, &nullifications, &mut sent, &mut sender).await;
-                            outstanding = Some((notarizations, nullifications, sent, status));
+                            self.send(Source::Required, false, &mut sender).await;
```

Comment:
With this + timeout + invalid handling, we may concurrently fetch the same notarizations/nullifications at the same time.